### PR TITLE
feat(rds): support setting database master users from existing secrets

### LINKED ISSUE #7927: Support SecretValue for DatabaseCluster masterUser.username
When we used Cloudformation directly, we used to set the master **username** and **password** for our DB cluster like this:

```json
...,
"MasterUsername":{"Fn::Sub":"{{resolve:secretsmanager:DB_CREDS:SecretString:username}}"},
"MasterUserPassword":{"Fn::Sub":"{{resolve:secretsmanager:DB_CREDS:SecretString:password}}"},
...
```
With the CDK, it seems like the credentials are defined as:

```ts
    const secret = new Secret(this, 'MyDBSecret', {
      secretName: 'DB_PASSWORD',
      generateSecretString: {
        excludePunctuation: true
      }
    });

    const dbCluster = new DatabaseCluster(this, 'MyDbCluster', {
      ...,
      masterUser: {
        username: 'myUserName',
        password: secret.secretValue
      },
      ...
    });
```
So only the **password** can be secure.

It would be nice to be able to set both the **username** and **password** based on a JSON secret value of the form:

```json
{
   "username": "myUserName",
   "password": "dontTellAnyone!"
}
```

I am currently using this workaround to achieve what I want:

```ts
    const username = SecretValue.secretsManager(DB_SECRET, { jsonField: 'username' }).toString();
    const password = SecretValue.secretsManager(DB_SECRET, { jsonField: 'password' });

    const dbCluster = new DatabaseCluster(this, 'MyDbCluster', {
      ...,
      masterUser: {
        username
        password
      },
      ...
    });
```

### Use Case

For improved security.

### Proposed Solution

In the **Login** data type, define the **username** field as **string | SecretValue**. This might make it more obvious to others that they can also secure the **username** but, at the same time, keep this backwards compatible.

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by njlynch**
See https://github.com/aws/aws-cdk/issues/7927#issuecomment-694857345 for motivation and design.

The current way of specifying master user logins for `DatabaseInstance` and
`DatabaseCluster` is inconsistent between the two and introduces some awkward
usage when creating a login from an existing `Secret`.

This change converts the existing `Login` interface (used by the `DatabaseCluster`)
into a class with factory methods for username/password or secret-based logins.
This also then re-uses that same interface for `DatabaseInstance`.

The one exception now will be `DatabaseInstanceFromSnapshot`, which has specific
requirements that deserved its own interface (`SnapshotLogin`).

As a side effect of this approach, existing `DatabaseCluster` users -- in
Typescript at least -- will not be broken. For example, the following are
equivalent:

```ts
new rds.DatabaseCluster(this, 'Cluster1', {
  // Existing usage
  masterUser: {
    username: 'admin',
  },
  // New usage
  masterUser: Login.fromUsername('admin'),
});
```

Lastly, this change makes the whole `masterUser` prop optional, as there's no good reason why we can't default a username.

fixes #7927

BREAKING CHANGE: `DatabaseInstanceProps` and `DatabaseInstanceFromSnapshotProps` -
`masterUsername`, `masterUserPassword` and `masterUserPasswordEncryptionKey` moved
to `credentials` as a new `Credentials` class.
* **rds:** `Login` renamed to `Credentials`. Use `Credentials.fromUsername` to replace existing usage.
* **rds:** `DatabaseClusterProps` `masterUser` renamed to `credentials`.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:425]: Gotta say, I'm not a huge fan of the `masterUser` name... How about we take this opportunity to rename it to something like `login`? (If you have a better name, I'm all ears 🙂)

Or, do our usual "deprecate + new field" dance if you want to minimize the breaking changes :)

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:887]: Same comment here as in Cluster - how about we come up with a better name for this concept, since we're changing it anyway?

File: packages/@aws-cdk/aws-rds/README.md
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/README.md:241]: i think it would be helpful to see a different usage of `Login` than just `fromUsername()` in the ReadMe.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:982]: Same here (rename to `login`?).

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/instance.ts:1004]: I feel like this error is rather cryptic. Will a user of `DatabaseInstanceFromSnapshot` understand what's gone wrong here, and know how to unblock themselves?

**Review by skinny85 — COMMENTED** [review]
Some initial comments.

I'm wondering what's the correct abstraction for `Login`. The edge case I'm thinking is when a customer needs to provide a custom implementation of `Login`, because ours doesn't cater to some edge-case. I guess with an abstract class, it might be tricky to subclass it from other languages, while a struct (+ class with static methods) should be easy... something to think about perhaps?

**Comment by njlynch** [comment]
> I'm wondering what's the correct abstraction for Login. 

I was wondering about this a bit myself. As-is, it's interesting that changing `Login` from an `interface` to `abstract class` was effectively a no-op for clusters (in Typescript). I like that the factory methods mean that you know that the combination of properties is always valid, but it's so easy to just anonymously subclass that there's nothing stopping anyone from continuing to use `Login` as it was already being used before (at least in some languages).

> I guess with an abstract class, it might be tricky to subclass it from other languages

Why would it be tricky to subclass it?

```java
Login login = new Login() {
  public String getUsername { return "Adam"; }
  public SecretValue getPassword { return myPassword; }
  // ...
}
```

> ... while a struct (+ class with static methods) should be easy.

Are you arguing for a `LoginFactory`? I don't know that I can bring myself to do that.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:425]: I do like `login` better, and was briefly tempted to do a rename, but held off as I liked that without that, there was a degree of backwards compatibility here. If users with clusters are able to keep their existing code -- and depending on language and specifics of how it's written, lots of users will be able to -- than that's a huge perk.

The "deprecate + new field" dance is an acceptable compromise though; I can do that throughout.

File: packages/@aws-cdk/aws-rds/README.md
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/README.md:241]: I agree, but I struggled to find the right place in the README to insert it. I'll give it another pass.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:1004]: What's cryptic about it? Either they used the factory methods, in which case this error is impossible, or they've manually created a Login object with `generatePassword` true and `username` blank. This error tells them to set the username. 

What's your suggestion for an alternative?

**Review by skinny85 — COMMENTED** [review]
> > I guess with an abstract class, it might be tricky to subclass it from other languages
> 
> Why would it be tricky to subclass it?
> 
> ```java
> Login login = new Login() {
>   public String getUsername { return "Adam"; }
>   public SecretValue getPassword { return myPassword; }
>   // ...
> }
> ```

I actually wasn't sure that even worked in Java (because of JSII), but I tried it, and it indeed does work.

> > ... while a struct (+ class with static methods) should be easy.
> 
> Are you arguing for a `LoginFactory`? I don't know that I can bring myself to do that.

The name could be cooler than that 🙂 For example, `Logins` sounds good to me.

I'm just saying that if `Login` was a struct, you could provide one like so:

```java
  Login login = Login.builder()
    .username(myUsername)
    .password(myPassword)
    .build();
```

Which I think is much more idiomatic in Java than an anonymous class.

But if you really really hate the idea of having 2 classes for `Login`, I won't block the PR on it. Just something to think about 🙂.

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (jogold) [inline packages/@aws-cdk/aws-rds/lib/props.ts:100]: I also prefer a concrete class here... with the static methods returning a `new Login()`... there's something similar here

https://github.com/aws/aws-cdk/blob/94bbabfff8b785656a40e85539f97ae8cb83c818/packages/%40aws-cdk/aws-amplify/lib/basic-auth.ts#L50-L53

I also wonder if `MasterCredentials` or just `Credentials` is a better name...

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (jogold) [inline packages/@aws-cdk/aws-rds/lib/props.ts:106]: Maybe move `password` and `encryptionKey` under an `options` object?

If a user wants to specify the username and generate a secret with a specific key he needs to do `fromUsername('superadmin', undefined, myKey)`...

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:982]: There is no `login` on `DatabaseInstanceSourceProps` -- there's one on `DatabaseInstanceProps` and one on `DatabaseInstanceFromSnapshotProps`, both of which extend from `DatabaseInstanceSourceProps`.

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:1017]: Agreed we can make it configurable later.

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/props.ts:100]: > I also prefer a concrete class here...

Out of curiosity, why? Just stylistic preference, or is there something else I'm missing?

I'm not strongly opposed; however, it's worth noting that changing `Login` from an interface (today) to a concrete class _will_ be a breaking change for all active Cluster users, meaning this PR will break *all* current RDS users (both Instance and Cluster setups). As an abstract class, that's not the case. If we're going to break, it does need to be now, but I'm not sure the minor aesthetics of the class being concrete are worth it.

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/props.ts:106]: Yeah, I was tempted to do this myself; good call.

**Comment by njlynch** [comment]
> > > ... while a struct (+ class with static methods) should be easy.
> 
> I'm just saying that if `Login` was a struct, you could provide one like so:
> 
> ```java
>   Login login = Login.builder()
>     .username(myUsername)
>     .password(myPassword)
>     .build();
> ```
> 
> Which I think is much more idiomatic in Java than an anonymous class.

Sure, but you only need to subclass (anonymously or otherwise) if you want to do something not supported by the factory methods; the factory methods support everything possible with the constructs, so I'm not sure what use case you're trying to optimize for here, and what the above offers that the below doesn't.

```java
Login login = Login.fromUsername(myUsername, myPassword);
```

If we want to optimize for the builder pattern, we could have chosen Option 2 from https://github.com/aws/aws-cdk/issues/7927#issuecomment-694857345, and skipped most of the breaking changes.

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (jogold) [reply packages/@aws-cdk/aws-rds/lib/props.ts:100]: If you go from this
```ts
interface Login {
  readonly username: string;
  readonly password?: string;
  readonly encryptionKey?: kms.IKey;
}
```

to this:
```ts
class Login {
  public static fromUsername(username: string, options: FromUsernameOptions = {}): Login {
    return new Login(username, options.password, options.encryptionKey);
  }

  public static fromSecret(secret: secretsmanager.Secret): Login {
    return new Login(
      secret.secretValueFromJson('username').toString(),
      secret.secretValueFromJson('username'),
      secret.encryptionKey,
      secret,
    );
  }

  public readonly username: string;
  public readonly password?: SecretValue;
  public readonly encryptionKey?: kms.IKey;

  constructor(username: string, password?: SecretValue, encryptionKey?: kms.IKey, secret?: secretsmanager.Secret) {
    this.username = username;
    this.password = password;
    this.encryptionKey = encryptionKey;
    this.secret = secret;
  }
}
```

it is not breaking, at least in TS, no?

Comment (njlynch) [reply packages/@aws-cdk/aws-rds/lib/props.ts:100]: Ah, you're right. I had done a slightly different implementation (getters) that didn't work; this approach is fine (at least in Typescript).

I'm still curious if the perceived advantage of this approach is stylistic, or if there's a benefit I'm missing.

**Review by skinny85 — APPROVED** [review]
Looks good to me!

One thing I would change is make the decision about concrete class vs. abstract once, and use it consistently between `Login` and `SnapshotLogin` - right now, `Login` is concrete, while `SnapshotLogin` is abstract.

I actually prefer the "struct + static factory" approach more than the concrete class (I don't like the 4 positional arguments in the constructor of the concrete class, and how it will translate to other languages; and if you make the constructor private, you're actually completely blocking the "escape hatch" of subclassing the class), so if the choice is between "concrete class" vs. "abstract class", I would go with "abstract class" here, but I'm deferring that decision to you, Nick 🙂.

File: packages/@aws-cdk/aws-rds/lib/props.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/props.ts]: +1 for `credentials` or `masterCredentials` as the name of the property, that's a great suggestion @jogold !

File: packages/@aws-cdk/aws-rds/lib/instance.ts
Comment (skinny85) [reply packages/@aws-cdk/aws-rds/lib/instance.ts:982]: Ah, you're right. All good then 🙂.

**Pull request merged**
